conans/client/loader.py

```diff
@@ -59,10 +59,16 @@ def load_basic_module(self, conanfile_path, lock_python_requires=None, user=None
             if self._pyreq_loader:
                 self._pyreq_loader.load_py_requires(conanfile, lock_python_requires, self)
 
-            conanfile.conan_data = self._load_data(conanfile_path)
-
-            self._cached_conanfile_classes[conanfile_path] = (conanfile, lock_python_requires,
-                                                              module)
+            # Load and populate dynamic fields from the data file
+            conan_data = self._load_data(conanfile_path)
+            conanfile.conan_data = conan_data
+            if conan_data and '.conan' in conan_data:
+                scm_data = conan_data['.conan'].get('scm_data')
+                if scm_data:
```

Comment:
This should ALWAYS be read from the `conandata.yml` because, regardless of the configuration in the user client consuming the recipe, if the packager has the opt-in activated and it is exporting the information in the `conandata.yml` file, the consumer has to read and use it.